src/Generating/EnderDragonFightStructuresGen.cpp

```diff
@@ -0,0 +1,272 @@
+				}
+				if (ShouldPlace)
+				{
+					m_TowerPos[Chunk].emplace_back(TowerProperties[I]);
+				}
+			}
+		}
+		// Generate angle of next tower
+		Angle = fmod(Angle + (2.0 * M_PI / static_cast<double>(TowerProperties.size())), 2.0 * M_PI);
+	}
+}
+
+
+
+
+
+void cEnderDragonFightStructuresGen::GenFinish(cChunkDesc &a_ChunkDesc)
+{
+	auto Coords = a_ChunkDesc.GetChunkCoords();
+	// If not in the chunks to write
+	if ((Coords.m_ChunkX > m_MaxX) ||
+		(Coords.m_ChunkX < m_MinX) ||
+		(Coords.m_ChunkZ > m_MaxZ) ||
+		(Coords.m_ChunkZ < m_MinZ))
+	{
+		return;
+	}
+	// Places the exit portal
+	if (Coords == cChunkCoords({0, 0}))
+	{
+		auto EnderDragon = std::make_unique<cEnderDragon>();
+		EnderDragon->SetPosition({0.0, static_cast<double>(a_ChunkDesc.GetHeight(0, 0) + 20), 0.0});  // Spawns the dragon 20 blocks above the terrain at (0, 0)
+		a_ChunkDesc.GetEntities().emplace_back(std::move(EnderDragon));
+		a_ChunkDesc.WriteBlockArea(m_Fountain,
+			static_cast<int>(FloorC(-m_Fountain.GetSizeX() / 2)),
+			62,
+			static_cast<int>(FloorC(-m_Fountain.GetSizeX() / 2)),
+			cBlockArea::msSpongePrint
+		);
+	}
+	else if (Coords == cChunkCoords({-1, 0}))
+	{
+		a_ChunkDesc.WriteBlockArea(m_Fountain,
+			cChunkDef::Width - static_cast<int>(FloorC(m_Fountain.GetSizeX() / 2)),
+			62,
+			static_cast<int>(FloorC(-m_Fountain.GetSizeZ() / 2)),
+			cBlockArea::msSpongePrint
+		);
+	}
+	else if (Coords == cChunkCoords({0, -1}))
+	{
+		a_ChunkDesc.WriteBlockArea(m_Fountain,
+			static_cast<int>(FloorC(-m_Fountain.GetSizeX() / 2)),
+			62,
+			cChunkDef::Width - static_cast<int>(FloorC(m_Fountain.GetSizeZ() / 2)),
+			cBlockArea::msSpongePrint);
+	}
+	else if (Coords == cChunkCoords({-1, -1}))
+	{
+		a_ChunkDesc.WriteBlockArea(m_Fountain,
+			cChunkDef::Width - static_cast<int>(FloorC(m_Fountain.GetSizeX() / 2)),
+			62,
+			cChunkDef::Width - static_cast<int>(FloorC(m_Fountain.GetSizeZ() / 2)),
+			cBlockArea::msSpongePrint);
+	}
+	auto It = m_TowerPos.find(Coords);
+	if (It == m_TowerPos.end())
+	{
+		return;
+	}
+	for (const auto & Property : It->second)
+	{
+		PlaceTower(a_ChunkDesc, Property);
+	}
+}
+
+
+
+
+
+void cEnderDragonFightStructuresGen::PlaceTower(cChunkDesc &a_ChunkDesc, const sTowerProperties & a_Properties)
+{
+	auto Pos = cChunk::AbsoluteToRelative(a_Properties.m_Pos, a_ChunkDesc.GetChunkCoords());
+
+	// Place obsidian pillar
+	for (int X = -a_Properties.m_Radius; X < a_Properties.m_Radius; X++)
+	{
+		for (int Z = -a_Properties.m_Radius; Z < a_Properties.m_Radius; Z++)
+		{
+			Vector3i NewPos = {Pos.x + X, 1, Pos.z + Z};
+			if (cChunkDef::IsValidRelPos(NewPos))
+			{
+				// The 3 was achieved by trial and error till the shape matched the notchian implementation
+				if ((NewPos - Vector3i(Pos.x, 1, Pos.z)).SqrLength() < a_Properties.m_Radius * a_Properties.m_Radius - 3)
+				{
+					for (int Y = 0; Y <= a_Properties.m_Height - 2; Y++)
+					{
+						a_ChunkDesc.SetBlockType(NewPos.x, Y, NewPos.z, E_BLOCK_OBSIDIAN);
```

Comment:
Maybe we should check for malformated towers to prevent a player from generating tower that surpass the height limit